### PR TITLE
Improve ctc loss

### LINKED ISSUE #4248: Optimize CTC loss
In the current CTC implementation, it uses a `(N, N)` dense connection matrix where `N` is length of a path. To calculates the next probability, each node requires `O(N^2)` time.  Each node only connects to at most three node (the same character, the previous character and the previous blank character. all weights for other nodes are filled with `-infity`). So its time complexity can be `O(N)`. We can reduce computational complexity drastically.

### PR DESCRIPTION
**Pull request by unnonouno**
I enhanced CTC loss function. Main difference is that in this fix I don't use recurrence connection matrix `frr` nor `brr` and instead access probability vector directory.
Please check two commits one by one. The first commit changed implementation strategy as I wrote. GPU performance degrades with the first one because it makes a lot of CUDA kernels. In the second commit, I reduced the number of kernels. Finally I only uses one elementwise kernel.
related to #4248

### CODE REVIEW

**Comment by unnonouno** [comment]
In my local test, forward computation is 10 times faster with CPU, and 2 times faster with GPU.

File: chainer/functions/loss/ctc.py [hunk 1]
```diff
@@ -168,44 +143,94 @@ def label_probability(self, label_size, path, path_length,
                                           path.shape[1], ret[i])
         return ret
 
+    def _computes_transition(self, prev_prob, path, path_length):
+        xp = cuda.get_array_module(prev_prob)
+
+        if xp == numpy:
+            n_batch, max_path_length = path.shape
+            mat = xp.full((3, n_batch, max_path_length), self.zero_padding, 'f')
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: flake8
`.\chainer\functions\loss\ctc.py:151:80: E501 line too long (80 > 79 characters)`

**Comment by unnonouno** [comment]
fixed

**Comment by okuta** [comment]
LGTM!

**Pull request merged**